src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -164,6 +178,52 @@ RouterExecutorStart(QueryDesc *queryDesc, int eflags)
 
 
 /*
+ * ReacquireMetadataLocks re-acquires the metadata locks that are normally
+ * acquired during planning.
+ *
+ * If we are executing a prepared statement, then planning might have
+ * happened in a separate transaction and advisory locks are no longer
+ * held. If a shard is currently being repaired/copied/moved, then
+ * obtaining the locks will fail and this function throws an error to
+ * prevent executing a stale plan.
+ *
+ * If we are executing a non-prepared statement or planning happened in
+ * the same transaction, then we already have the locks and obtain them
+ * again here. Since we always release these locks at the end of the
+ * transaction, this is effectively a no-op.
+ */
+static void
+ReacquireMetadataLocks(List *taskList)
+{
+	ListCell *taskCell = NULL;
+
+	/*
+	 * Note: to avoid the overhead of additional sorting, we assume tasks
+	 * to be already sorted by shard ID such that deadlocks are avoided.
+	 * This is true for INSERT/SELECT, which is the only multi-shard
+	 * command right now.
+	 */
+
+	foreach(taskCell, taskList)
+	{
+		Task *task = (Task *) lfirst(taskCell);
+
+		/*
+		 * Only obtain metadata locks for modifications to allow reads to
+		 * proceed during shard copy.
+		 */
+		if (task->taskType == MODIFY_TASK &&
+			!TryLockShardDistributionMetadata(task->anchorShardId, ShareLock))
+		{
+			ereport(ERROR, (errcode(ERRCODE_LOCK_NOT_AVAILABLE),
```

Comment:
I guess that was in reference to the serializable comment, not the odd locking pattern I was pondering?